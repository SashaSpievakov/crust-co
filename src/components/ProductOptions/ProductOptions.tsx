import { FC, KeyboardEvent, useLayoutEffect } from 'react';

import { Li, Wrapper } from './ProductOptions.styled';
import { IProductOptions } from './ProductOptions.type';

export const ProductOptions: FC<IProductOptions> = ({
  price,
  sizes,
  types,
  activeSize,
  activeType,
  setActivePrice,
  setActiveSize,
  setActiveType,
  typeNames,
  isFullScreen = false,
}) => {
  const typeClickHandler = (newType: number) => {
    if (activeType !== newType) {
      setActivePrice(activeType < newType ? price + 1 : price - 1);
    }

    setActiveType(newType);
  };

  const sizeClickHandler = (newSize: number) => {
    const priceDifference = newSize - activeSize;
    setActivePrice(price + priceDifference);

    setActiveSize(newSize);
  };

  const handleKeyDownType = (e: KeyboardEvent<HTMLElement>, type: number) => {
    if (e.code === 'Enter') typeClickHandler(type);
  };

  const handleKeyDownSize = (e: KeyboardEvent<HTMLElement>, size: number) => {
    if (e.code === 'Enter') sizeClickHandler(size);
  };

  useLayoutEffect(() => {
    setActiveSize(sizes[0]);
    setActiveType(types[0]);
  }, [sizes]);

  return (
    <Wrapper isFullScreen={isFullScreen}>
      <ul>
        {types.map((type: number) => (
          <Li
            key={type}
            onClick={() => typeClickHandler(type)}
            onKeyDown={(e) => handleKeyDownType(e, type)}
            chosen={activeType === types.indexOf(type)}
            tabIndex={0}
            aria-current={activeType === types.indexOf(type)}
          >
            {typeNames[type]}
          </Li>
        ))}
      </ul>
      <ul>
        {sizes.map((size: number) => (
          <Li
            key={size}
            onClick={() => sizeClickHandler(size)}
            onKeyDown={(e) => handleKeyDownSize(e, size)}
            chosen={activeSize === size}
            tabIndex={0}
            aria-current={activeSize === size}
          >
            {size} inch
          </Li>
        ))}
      </ul>
    </Wrapper>
  );
};
