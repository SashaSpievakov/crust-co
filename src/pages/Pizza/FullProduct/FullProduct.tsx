import { FC, useState } from 'react';

import { ItemsCountHandler, ProductOptions } from '@src/components';
import { ButtonLink } from '@src/components/UI';
import { IPizzaItem } from '@src/models';

import { Image, Item, Paragraph, Price, Title } from './FullProduct.styled';

interface FullProductProps {
  item: IPizzaItem;
}

export const typeNames: string[] = ['traditional', 'thin'];

export const FullProduct: FC<FullProductProps> = ({ item }) => {
  const [activePrice, setActivePrice] = useState<number>(item.price);
  const [activeSize, setActiveSize] = useState<number>(12);
  const [activeType, setActiveType] = useState<number>(0);

  return (
    <Item data-testid="itemPage">
      <picture>
        <source
          srcSet={`../assets/img/pizzas/pizza${item.id}.webp`}
          type="image/webp"
        />
        <source
          srcSet={`../assets/img/pizzas/pizza${item.id}.png`}
          type="image/png"
        />
        <Image
          src={`../assets/img/pizzas/pizza${item.id}.png`}
          alt={`${item.name} pizza`}
          width="450"
          height="450"
        />
      </picture>
      <Title>{item.name}</Title>
      <Paragraph>{item.description}</Paragraph>
      <ProductOptions
        price={activePrice}
        sizes={item.sizes}
        types={item.types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
        isFullScreen
      />
      <Price>{activePrice}$</Price>
      <ItemsCountHandler
        id={item.id}
        name={item.name}
        price={activePrice}
        activeSize={activeSize}
        activeType={activeType}
        typeNames={typeNames}
        isFullScreen
      />
      <ButtonLink link="/">Go Back</ButtonLink>
    </Item>
  );
};
