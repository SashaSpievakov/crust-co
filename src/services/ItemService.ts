import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

const itemAPI = createApi({
  reducerPath: 'itemAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Item'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (id) => ({
        url: `/pizzas/${id}`,
      }),
      providesTags: () => ['Item'],
    }),
  }),
});

export default itemAPI;
