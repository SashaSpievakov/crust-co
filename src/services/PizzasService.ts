import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

interface IFetchItems {
  activeCategory: number;
  sortedPropertyName: string;
}

const pizzasAPI = createApi({
  reducerPath: 'pizzasAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: 'https://6344adb1dcae733e8fe3067a.mockapi.io/',
  }),
  tagTypes: ['Pizzas'],
  endpoints: (build) => ({
    fetchPizzas: build.query<IPizzaItem[], IFetchItems>({
      query: ({ activeCategory, sortedPropertyName }) => ({
        url: `pizza-items?${
          activeCategory > 0 ? `category=${activeCategory}&` : ''
        }sortBy=${sortedPropertyName}`,
      }),
      providesTags: () => ['Pizzas'],
    }),
  }),
});

export default pizzasAPI;
