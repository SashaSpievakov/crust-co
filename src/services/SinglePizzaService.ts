import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

export const pizzaAPI = createApi({
  reducerPath: 'pizzaPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Pizza'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (name) => ({
        url: `/pizzas/${name}`,
      }),
      providesTags: (result) =>
        result ? [{ type: 'Pizza', id: result.name }] : ['Pizza'],
    }),
  }),
});
