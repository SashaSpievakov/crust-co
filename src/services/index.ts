export { pizzasAPI } from './PizzasService';
export { pizzaAPI } from './SinglePizzaService';
