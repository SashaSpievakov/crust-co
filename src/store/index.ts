export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
