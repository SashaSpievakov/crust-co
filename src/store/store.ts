import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
