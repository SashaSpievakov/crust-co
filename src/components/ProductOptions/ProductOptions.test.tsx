import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { typeNames } from '../../pages/Pizza/FullProduct/FullProduct';
import { pizzaAPI } from '../../services';
import { setupStore } from '../../store';
import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '../../tests/helpers';
import { testServer } from '../../tests/mocks';
import { ProductOptions } from './ProductOptions';
import { IProductOptions } from './ProductOptions.type';

const ProductOptionsMockProps: IProductOptions = {
  price: 10,
  sizes: [12, 14, 16],
  types: [0, 1],
  activeSize: 0,
  activeType: 0,
  setActivePrice: jest.fn(),
  setActiveSize: jest.fn(),
  setActiveType: jest.fn(),
  typeNames,
  isFullScreen: true,
};

describe('ProductOptions tests', () => {
  test('should render the ProductOptions UI component', () => {
    const snapshot = rendererWithProviders(
      <ProductOptions {...ProductOptionsMockProps} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('ProductOptions select logic', () => {
    beforeAll(() => {
      testServer.listen();
    });

    afterAll(() => {
      testServer.close();
    });

    beforeEach(() => {
      renderWithProvidersAndRoutes(null, true, '/pizzas/Chicken%20Curry');
    });

    afterEach(() => {
      testServer.resetHandlers();
      setupStore().dispatch(pizzaAPI.util.resetApiState());
    });

    test('should check thin type selection', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const typeThin = screen.getByText(/thin/i);

      userEvent.click(typeThin);

      expect(price).toHaveTextContent(/15\$/i);
    });

    test('should change back to traditional type', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const typeTraditional = screen.getByText(/traditional/i);
      const typeThin = screen.getByText(/thin/i);

      userEvent.click(typeThin);
      userEvent.click(typeTraditional);

      expect(price).toHaveTextContent(/14\$/i);
    });

    test('should check double thin type selection', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const typeThin = screen.getByText(/thin/i);

      userEvent.click(typeThin);
      userEvent.click(typeThin);

      expect(price).toHaveTextContent(/15\$/i);
    });

    test('should change from 12 inch to 14 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const fourteen = screen.getByText(/14 inch/i);

      userEvent.click(fourteen);

      expect(price).toHaveTextContent(/16\$/i);
    });

    test('should change from 12 inch to 16 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(sixteen);

      expect(price).toHaveTextContent(/18\$/i);
    });

    test('should change from 14 inch to 16 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const fourteen = screen.getByText(/14 inch/i);
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(fourteen);
      expect(price).toHaveTextContent(/16\$/i);

      userEvent.click(sixteen);
      expect(price).toHaveTextContent(/18\$/i);
    });

    test('should change from 14 inch to 12 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const fourteen = screen.getByText(/14 inch/i);
      const twelve = screen.getByText(/12 inch/i);

      userEvent.click(fourteen);
      expect(price).toHaveTextContent(/16\$/i);

      userEvent.click(twelve);
      expect(price).toHaveTextContent(/14\$/i);
    });

    test('should change from 16 inch to 14 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const fourteen = screen.getByText(/14 inch/i);
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(sixteen);
      expect(price).toHaveTextContent(/18\$/i);

      userEvent.click(fourteen);
      expect(price).toHaveTextContent(/16\$/i);
    });

    test('should change from 16 inch to 12 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const twelve = screen.getByText(/12 inch/i);
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(sixteen);
      expect(price).toHaveTextContent(/18\$/i);

      userEvent.click(twelve);
      expect(price).toHaveTextContent(/14\$/i);
    });

    test('should select thin type and 16 inch together', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const typeThin = screen.getByText(/thin/i);
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(sixteen);
      userEvent.click(typeThin);

      expect(price).toHaveTextContent(/19\$/i);
    });

    test('should check triple and double click on 14 inch', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const fourteen = screen.getByText(/14 inch/i);

      userEvent.click(fourteen);
      userEvent.click(fourteen);
      expect(price).toHaveTextContent(/16\$/i);

      userEvent.click(fourteen);
      expect(price).toHaveTextContent(/16\$/i);
    });

    test('should check all clicks combined', async () => {
      const price = await screen.findByRole('heading', {
        level: 3,
      });
      const traditionalType = screen.getByText(/traditional/i);
      const thinType = screen.getByText(/thin/i);
      const twelve = screen.getByText(/12 inch/i);
      const fourteen = screen.getByText(/14 inch/i);
      const sixteen = screen.getByText(/16 inch/i);

      userEvent.click(sixteen);
      userEvent.click(traditionalType);
      userEvent.click(fourteen);
      userEvent.click(twelve);
      userEvent.click(thinType);
      userEvent.click(fourteen);
      userEvent.click(traditionalType);
      userEvent.click(sixteen);
      userEvent.click(twelve);
      userEvent.click(thinType);

      expect(price).toHaveTextContent(/15\$/i);
    });
  });
});
