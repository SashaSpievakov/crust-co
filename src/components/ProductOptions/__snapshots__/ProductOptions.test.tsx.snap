// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductOptions tests should render the ProductOptions UI component 1`] = `
Object {
  "children": Array [
    <ul>
      <li
        aria-current={true}
        className="sc-fEXmlR eEKhPW"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        traditional
      </li>
      <li
        aria-current={false}
        className="sc-fEXmlR jmBAcN"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        thin
      </li>
    </ul>,
    <ul>
      <li
        aria-current={false}
        className="sc-fEXmlR jmBAcN"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        12
         inch
      </li>
      <li
        aria-current={false}
        className="sc-fEXmlR jmBAcN"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        14
         inch
      </li>
      <li
        aria-current={false}
        className="sc-fEXmlR jmBAcN"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        16
         inch
      </li>
    </ul>,
  ],
  "props": Object {
    "className": "sc-fnGiBr gOOXPX",
  },
  "type": "div",
}
`;
